tools/rws_backup.sh

```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
+set -eu
+
+
+# The directory where the mount point will be
+# created. Used for cleanup.
+WORK_DIR=$(pwd)
+
+
+help() {
+	cat <<EOF
+rws_backup.sh - util for working with S3 backups.
+
+Usage:
+	rws_backup.sh COMMAND
+
+Commands:
+	help        Show help.
+	backup      Create a backup.
+
+To specify parameters the environment variables are used.
+
+Parameters:
+	AWSACCESSKEYID      Access Key ID for the account.
+	AWSSECRETACCESSKEY  Secret Access Key for the account.
+	RESTIC_PASSWORD     Restic repository password.
+	RWS_BACKUP_URL      URL to use to access S3 (example: https://hb.bizmrg.com).
+	RWS_BACKUP_BUCKET   Backed up bucket name.
+	RWS_BACKUP_PATH     Path inside the backed up bucket (if want to back up
+	                    only a part of the bucket).
+	RWS_BACKUP_DEST     Path to save the backup (examples: /path/to/backup ,
+	                    s3:hb.bizmrg.com/backups-bucket/backup). When using S3
+	                    as the backup destination, the same AWS Access key will
+	                    be used for both the backup source bucket and the backup
+	                    destination bucket.
+EOF
+}
+
+
+cleanup() {
+	# Cleanup after complete backup.
+	cd ${WORK_DIR}
+	# We use "set -e" but we don't want to exit if the "unmount"
+	# command fails, because a temporary directory may has been
+	# created and some error occurred while trying to mount.
+	# Therefore, we use "|| true".
+	umount ${S3_MOUNT_POINT} || true
+	rmdir ${S3_MOUNT_POINT}
+}
+
+
+prepare_restic_repo() {
+	# Check if the repository exists
+	# (see https://github.com/restic/restic/issues/1690).
+	# Init the repository if it doesn't exist or exit if the
+	# password is wrong.
+	local CHECK_RESULT=0
+	local CHECK_OUT=$(restic --repo ${RWS_BACKUP_DEST} cat config 2>&1) \
+		|| CHECK_RESULT=${?}
+
+	if [[ ${CHECK_RESULT} != 0 ]]; then
+		if [[ ${CHECK_OUT} == "Fatal: wrong password or no key found" ]]; then
+			echo ${CHECK_OUT}
+			exit 1
+		else
+			restic init --repo ${RWS_BACKUP_DEST}
```

Comment:
Shouldn't we do it only when a specific error message was reported by restic? Not in any situation, when the error message is not 'Fatal: wrong password or no key found'.

Whether this action is harmless for an existing repository? Does not it delete files in the case?

Reply:
Hm. It seems, I misunderstood what is going on. It is local from the point of view of restic. Just initialize a backup directory in a restic specific format, right?

So, maybe, we can check, whether the directory is in this format: say, look, whether a file with some format specific name exists?

Reply:
Looks relevant: https://restic.readthedocs.io/en/stable/075_scripting.html#check-if-a-repository-is-already-initialized

Reply:
> Looks relevant: https://restic.readthedocs.io/en/stable/075_scripting.html#check-if-a-repository-is-already-initialized

https://github.com/restic/restic/issues/1690#issuecomment-620197338 Seems like `restic cat config` is faster.

Reply:
> Hm. It seems, I misunderstood what is going on. It is local from the point of view of restic. Just initialize a backup directory in a restic specific format, right?
> 

Not necessarily local, maybe S3.

> So, maybe, we can check, whether the directory is in this format: say, look, whether a file with some format specific name exists?

This is what I am doing.

Reply:
> Shouldn't we do it only when a specific error message was reported by restic? Not in any situation, when the error message is not 'Fatal: wrong password or no key found'.
> 

That sounds good. I'll update this part of the code.

> Whether this action is harmless for an existing repository? Does not it delete files in the case?

As far as I see - No.